### PR TITLE
refactor: use unit testable real python functions for most templated remote_functions code

### PR DESCRIPTION
**Pull request by tswast**
Thank you for opening a Pull Request! Before submitting your PR, there are a few things you can do to make sure it goes smoothly:
- [ ] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery-dataframes/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [ ] Ensure the tests and linter pass
- [ ] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes #<issue_number_goes_here> 🦕

### CODE REVIEW

File: bigframes/dtypes.py [hunk 1]
```diff
@@ -755,10 +755,11 @@ def infer_literal_arrow_type(literal) -> typing.Optional[pa.DataType]:
 # Input and output types supported by BigQuery DataFrames remote functions.
 # TODO(shobs): Extend the support to all types supported by BQ remote functions
 # https://cloud.google.com/bigquery/docs/remote-functions#limitations
-SUPPORTED_IO_PYTHON_TYPES = {bool, float, int, str}
+SUPPORTED_IO_PYTHON_TYPES = {bool, bytes, float, int, str}
```
Comment (tswast) [inline bigframes/dtypes.py:758]: TODO: revert these changes. Save them for a future PR.

File: tests/system/small/test_remote_function.py [hunk 2]
```diff
@@ -483,18 +483,20 @@ def add_one(x):
     assert_pandas_df_equal(bf_result, pd_result)
 
 
-@pytest.mark.flaky(retries=2, delay=120)
+# @pytest.mark.flaky(retries=2, delay=120)
```
Comment (tswast) [inline tests/system/small/test_remote_function.py:486]: Revert these changes too.

File: tests/unit/functions/test_remote_function_template.py [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+    (
+        pytest.param(
+            json.dumps(
+                {
+                    "names": ["'my-index'", "'col1'", "'col2'", "'col3'"],
+                    "types": ["string", "Int64", "Int64", "Int64"],
+                    "values": ["my-index-value", "1", None, "-1"],
+                    # TODO: Why do we even have this? Does it matter if a
+                    # column was an index column or not? Looks like the index
+                    # columns get completely discarded, anyway?
```
Comment (tswast) [inline tests/unit/functions/test_remote_function_template.py:35]:
```suggestion
```

index value gets used as the name of the series.

File: bigframes/functions/remote_function_template.py [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+
+    # original code, only for debugging purpose
+    udf_code = textwrap.dedent(inspect.getsource(def_))
+    udf_code_file_path = os.path.join(directory, udf_code_file_name)
+    with open(udf_code_file_path, "w") as f:
+        f.write(udf_code)
+
+    # serialized bytecode
+    udf_bytecode_file_path = os.path.join(directory, udf_bytecode_file_name)
+    with open(udf_bytecode_file_path, "wb") as f:
```
Comment (GarrettWu) [inline bigframes/functions/remote_function_template.py:186]: Do we really need files on disk or File Objects in memory are enough? Potentially better performance.

File: tests/unit/functions/test_remote_function_template.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+                dtype="Int64",
+                index=["col1", "col2", "col3"],
+                name="my-index-value",
+            ),
+            id="int64-string-index",
+        ),
+        # TODO: test more types and indexes
+    ),
+)
+def test_get_pd_series(row_json, expected):
```
Comment (GarrettWu) [inline tests/unit/functions/test_remote_function_template.py:48]: Is this the only reason we want to extract get_pd_series? Do we want to add more unit tests for other functions?

**Review by GarrettWu — APPROVED** [review]

File: bigframes/functions/remote_function_template.py (hunk 4 above)
Comment (tswast) [reply bigframes/functions/remote_function_template.py:186]: I can try this out in a future PR. I'd like to keep this a pure refactor without any behavior changes. Filed issue 345433300.

```suggestion
    # TODO(b/345433300): try io.BytesIO to avoid writing to the file system
    with open(udf_bytecode_file_path, "wb") as f:
```
